File: app/src/main/java/eu/michaelvogt/ar/author/data/VisualDetail.java
/*
    ARTester - AR for tourists by tourists
    Copyright (C) 2018  Michael Vogt

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

package eu.michaelvogt.ar.author.data;

import android.arch.persistence.room.ColumnInfo;
import android.arch.persistence.room.Entity;
import android.arch.persistence.room.ForeignKey;
import android.arch.persistence.room.PrimaryKey;

import com.google.ar.sceneform.math.Quaternion;
import com.google.ar.sceneform.math.Vector3;

import eu.michaelvogt.ar.author.utils.Converter;

@Entity(tableName = "visual_detail",
    foreignKeys = @ForeignKey(entity = Area.class,
        parentColumns = "u_id",
        childColumns = "area_id"))
public class VisualDetail {
  private static final String TAG = VisualDetail.class.getSimpleName();

  // Default detail
  public static final int KEY_COORDTYPE = 28;
  public static final int KEY_POSITION = 29;
  public static final int KEY_ZEROPOINT = 30;
  public static final int KEY_ROTATION = 32;
  public static final int KEY_SIZE = 33;
  public static final int KEY_SCALE = 34;

  // Optional detail
  public static final int KEY_IMAGEPATH = 0;
  public static final int KEY_IMAGEFOLDERPATH = 1;
  public static final int KEY_TEXTCONTENT = 2;
  public static final int KEY_MARKUPCONTENT = 3;
  public static final int KEY_TEXTPATH = 4;
  public static final int KEY_SLIDES = 5;
  public static final int KEY_FADE_LEFT_WIDTH = 6;
  public static final int KEY_FADE_RIGHT_WIDTH = 7;
  public static final int KEY_FADE_TOP_WIDTH = 8;
  public static final int KEY_FADE_BOTTOM_WIDTH = 9;
  public static final int KEY_BACKGROUNDCOLOR = 10;
  public static final int KEY_TEXTCOLOR = 11;
  public static final int KEY_TEXTSIZE = 12;
  public static final int KEY_ALLOWZOOM = 13;
  public static final int KEY_IMAGERESOURCE = 14;
  public static final int KEY_ISCASTINGSHADOW = 15;
  public static final int KEY_HTMLPATH = 18;
  public static final int KEY_ZOOMINSIZE = 19;
  public static final int KEY_ZOOMINPOSITION = 20;
  public static final int SECONDARYTEXTURE = 21;
  public static final int KEY_SECONDARYIMAGEPATH = 22;
  public static final int KEY_TITLE = 23;
  public static final int KEY_RESOURCE = 24;
  public static final int KEY_LANGUAGE = 25;
  public static final int KEY_SCALEVALUES = 26;
  public static final int KEY_IMAGEDESCRIPTIONS = 27;

  public static final int TYPE_ALL = 0;

  @PrimaryKey(autoGenerate = true)
  @ColumnInfo(name = "u_id")
  private int uId;

  @ColumnInfo(name = "area_id")
  private int areaId;

  @ColumnInfo(name = "type")
  private int type;

  @ColumnInfo(name = "target")
  private int target;

  @ColumnInfo(name = "value")
  public String value;

  public VisualDetail() {
    this(0, 0, 0, 0);
  }

  public VisualDetail(int areaId, int type, int target, Object value) {
    this.areaId = areaId;
    this.type = type;
    this.target = target;
    this.value = Converter.stringify(value);
  }

  public int getUId() {
    return uId;
  }

  public void setUId(int uId) {
    this.uId = uId;
  }

  public int getAreaId() {
    return areaId;
  }

  public void setAreaId(int areaId) {
    this.areaId = areaId;
  }

  public int getType() {
    return type;
  }

  public void setType(int type) {
    this.type = type;
  }

  public int getTarget() {
    return target;
  }

  public void setTarget(int target) {
    this.target = target;
  }

  public Object getValue() {
    return Converter.objectify(value);
  }

  public void setValue(Object value) {
    this.value = Converter.stringify(value);
  }
}

File: app/src/main/java/eu/michaelvogt/ar/author/data/VisualDetail.kt
/*
    ARTester - AR for tourists by tourists
    Copyright (C) 2018  Michael Vogt

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

package eu.michaelvogt.ar.author.data

import androidx.room.*
import eu.michaelvogt.ar.author.data.utils.Converters

private val TAG = VisualDetail::class.java.simpleName

const val KEY_IMAGEPATH = 0
const val KEY_IMAGEFOLDERPATH = 1
const val KEY_TEXTCONTENT = 2
const val KEY_MARKUPCONTENT = 3
const val KEY_TEXTPATH = 4
const val KEY_SLIDES = 5
const val KEY_FADE_LEFT_WIDTH = 6
const val KEY_FADE_RIGHT_WIDTH = 7
const val KEY_FADE_TOP_WIDTH = 8
const val KEY_FADE_BOTTOM_WIDTH = 9
const val KEY_BACKGROUNDCOLOR = 10
const val KEY_TEXTCOLOR = 11
const val KEY_TEXTSIZE = 12
const val KEY_ALLOWZOOM = 13
const val KEY_IMAGERESOURCE = 14
const val KEY_ISCASTINGSHADOW = 15
const val KEY_HTMLPATH = 18
const val KEY_ZOOMINSIZE = 19
const val KEY_ZOOMINPOSITION = 20
const val KEY_SECONDARYTEXTURE = 21
const val KEY_SECONDARYIMAGEPATH = 22
const val KEY_TITLE = 23
const val KEY_RESOURCE = 24
const val KEY_LANGUAGE = 25
const val KEY_SCALEVALUES = 26
const val KEY_IMAGEDESCRIPTIONS = 27

const val TYPE_DETAIL_ALL = 0

@Entity(
        tableName = "visual_details",
        indices = [
            Index(value = ["area_id"])
        ])
class VisualDetail(
        @field:ColumnInfo(name = "area_id")
        var areaId: Long = 0,

        @field:ColumnInfo(name = "type")
        var type: Int = 0,

        @field:ColumnInfo(name = "key")
        var key: Int = -1,

        value: Any = "") {
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "u_id")
    var uId: Long = 0

    @ColumnInfo(name = "value")
    protected var value: String = ""

    @Ignore
    var slides: List<Slide> = emptyList()

    var anyValue: Any
        get() = Converters().objectify(value)
        set(anyValue) {
            value = Converters().stringify(anyValue)
        }

    init {
        this.value = Converters().stringify(value)
    }
}
